pcdsdevices/valve.py

```diff
@@ -0,0 +1,160 @@
+"""
+Standard classes for LCLS Gate Valves
+"""
+import logging
+from enum import Enum
+
+from ophyd import EpicsSignal, EpicsSignalRO, Component as C
+
+from .inout import InOutPositioner, InOutPVStatePositioner
+
+logger = logging.getLogger(__name__)
+
+
+class Commands(Enum):
+    """
+    Command aliases for opening and closing valves
+    """
+    close_valve = 0
+    open_valve = 1
+
+
+class InterlockError(PermissionError):
+    """
+    Error when request is blocked by interlock logic
+    """
+    pass
+
+
+class Stopper(InOutPVStatePositioner):
+    """
+    Controls Stopper
+
+    A base class for a device with two limits switches controlled via an
+    external command PV. This full encompasses the controls Stopper
+    installations as well as un-interlocked GateValves
+
+    Parameters
+    ----------
+    prefix : str
+        Full PPS Stopper PV
+
+    name : str, optional
+        Alias for the stopper
+
+    Attributes
+    ----------
+    commands : Enum
+        An enum with integer values for open_valve, close_valve values
+    """
+    # Default attributes
+    _default_read_attrs = ['open_limit', 'closed_limit']
+
+    # Limit-based states
+    open_limit = C(EpicsSignalRO, ':OPEN')
+    closed_limit = C(EpicsSignalRO, ':CLOSE')
+
+    # Information on device control
+    command = C(EpicsSignal, ':CMD')
+    commands = Commands
+
+    _state_logic = {'open_limit': {0: 'defer',
+                                   1: 'OUT'},
+                    'closed_limit': {0: 'defer',
+                                     1: 'IN'}}
+
+    def _do_move(self, state):
+        if state.name == 'IN':
+            self.command.put(self.commands.close_valve.value)
+        elif state.name == 'OUT':
+            self.command.put(self.commands.open_valve.value)
+
+    def open(self, **kwargs):
+        """
+        Open the stopper
+        """
+        return self.remove(**kwargs)
+
+    def close(self, **kwargs):
+        """
+        Close the stopper
+        """
+        return self.insert(**kwargs)
+
+
+class GateValve(Stopper):
```

Comment:
My non-vacuum brain finds this inheritance path very confusing, since stoppers block the beam robustly and I've heard stories of blowing through gate valves. It works in python because they behave basically the same, but I can't help but feel like there's a less confusing way to structure the OOP.

Reply:
I realize it doesn't make sense physically, but it make sense inheritance-wise. To EPICS, a GateValve is just a Stopper with more restrictions. I don't really want to add a third class but I can if you want. `Stopper` can not inherit from `GateValve` because you would need to delete the interlock logic.

The OOP would then be `Valve`, which is then subclasses by `GateValve` and `Stopper`?

Reply:
OOP would be something more abstract like `OpenCloseLimitsDevice` (ok, maybe a shorter name) that feeds into both `GateValve` and `Stopper`. This is not important to do, unless later we decide that `Stopper` needs a feature that `GateValve` should definitely not have. We can of course just refactor if the need arises.